### PR TITLE
Add a healthcheck endpoint on the ingesters that distributors can use

### PR DESCRIPTION
**Pull request by csmarchbanks**
This puts a Check endpoint into the ingesters that allows our ingester clients to do a healthcheck loop to make sure their connection is still ok. As part of the maintenance loop to remove stale ingesters, we will now also (in the background) healthcheck each remaining ingester, and if it is not ok delete the entry in distributor.clients. The client will then be remade the next time it is needed by `getClientFor`. 

A couple questions:
1. As it stands the ingesters need to be upgraded before the ruler/distributors/querier, or else the healtchecks will constantly fail and we will be recreating a lot of grpc clients. Everything will still work, but the logs would be noisy/extra work will be done. Is that a concern for those with CI systems? If so how should it be addressed?
2. The Check endpoint is based off https://github.com/grpc/grpc/blob/master/doc/health-checking.md. Is that what we want? Or should we make an endpoint that is more ingester specific?

Another solution would be to just periodically close the ingester clients, perhaps put an age field on them and only use them until then.

This PR fixes: https://github.com/weaveworks/cortex/issues/702

Current state of this PR:
* Ingesters have a new `Check` grpc endpoint, used to health check them
* Refactored ingester clients in a distributor to be their own struct, `IngesterPool`
* The `IngesterPool` has a function, `CleanUnhealthy()`, that will clean up unhealthy ingester clients
* `CleanUnhealthy()` is run as part of the distributor periodic cleanup when the `distributor.health-check-ingesters` flag is turned on
* util/compat.go was moved to ingester/client/compat.go in order to avoid cyclic dependencies

### CODE REVIEW

File: pkg/ingester/client/cortex.proto [hunk 1]
```diff
@@ -122,3 +123,16 @@ message LabelMatcher {
   string name = 2;
   string value = 3;
 }
+
+message HealthCheckRequest {
+  string service = 1;
+}
+
+message HealthCheckResponse {
+  enum ServingStatus {
```
Comment (limscoder) [inline pkg/ingester/client/cortex.proto]: Do UNKNOWN or NOT_SERVING ever get sent?

Comment (csmarchbanks) [reply pkg/ingester/client/cortex.proto]: They do not right now, that is the format suggested for health-checking by the gRPC docs. We could make a more cortex specific format if we want.

Comment (csmarchbanks) [reply pkg/ingester/client/cortex.proto]: It will also be compatible with: https://godoc.org/google.golang.org/grpc/health

I could also try to spend some time importing the proto definitions from grpc/health, rather than duplicating them.

**Comment by csmarchbanks** [comment]
This has been running for more than a day in staging, and is now running in our prod cluster. 

Started restarting gRPC clients ~17:00 on 3/10 
screen shot 2018-03-12 at 11 09 49 am

**Comment by tomwilkie** [comment]
> As it stands the ingesters need to be upgraded before the ruler/distributors/querier

General approach to this is to hide the behaviour behind a flag that defaults to off; new code gets rolled out with the behaviour off, then a flag changed can be rolled out to enable it.

File: pkg/distributor/distributor.go [hunk 2]
```diff
@@ -184,29 +184,55 @@ func (d *Distributor) Stop() {
...
+	}
+}
+
+func (d *Distributor) healthCheckAndRemoveIngester(addr string, client client.IngesterClient, wg *sync.WaitGroup) {
+	level.Debug(util.Logger).Log("msg", "healthchecking ingester client", "addr", addr)
+
+	ctx, cancel := context.WithTimeout(context.Background(), d.cfg.RemoteTimeout)
+	ctx = user.InjectOrgID(ctx, "0")
+	resp, err := client.Check(ctx, &ingester_client.HealthCheckRequest{})
+	cancel()
```
Comment (tomwilkie) [inline pkg/distributor/distributor.go]: Nit: can we move that up to the `ctx, cancel := ...` line and make it a defer please?  Just causes minor confusion as it doesn't match the common pattern.

File: pkg/distributor/distributor.go [hunk 3]
```diff
@@ -184,29 +184,55 @@ func (d *Distributor) Stop() {
 }
 
 func (d *Distributor) removeStaleIngesterClients() {
-	d.clientsMtx.Lock()
-	defer d.clientsMtx.Unlock()
```
Comment (tomwilkie) [inline pkg/distributor/distributor.go]: :-( I prefer the defer way, mainly because it makes the lock more robust to future modifications.

As the remote timeout is set to 2s, and the removeStateIngesterClients functions only runs every 15s, do we really need the extra goroutines and waitgroup?

File: pkg/ingester/client/cortex.proto (hunk 1 above)
Comment (tomwilkie) [reply pkg/ingester/client/cortex.proto]: Yeah, I'd say either use those protos or make our response an empty response.

**Comment by tomwilkie** [comment]
> From staging, started restarting gRPC clients ~17:00 on 3/10

Oh dear - I'm fine with fixes like this going in work around other issues, but do you have any idea why gRPC is getting so unwell?

**Comment by bboreham** [comment]
Does this help with #157 - use this endpoint as the heartbeat instead of the consul timestamp?

**Comment by tomwilkie** [comment]
> Does this help with #157 - use this endpoint as the heartbeat instead of the consul timestamp?

Not as it stands (ingester clients will be created on demand), but it could be made to do so...

**Comment by csmarchbanks** [comment]
@tomwilkie Thanks for the comments!

I wish I knew why gRPC is getting so unwell, I have been banging my head against a wall for awhile about this. If you or someone else have any ideas I would be happy to test them out.

Also, do you think there is anything we could do in this PR to make it easier to heartbeat with respect to #157? I don't want to do all the work for that here, but would be happy to make it easier in the future.

File: pkg/distributor/distributor.go (hunk 3 above)
Comment (csmarchbanks) [reply pkg/distributor/distributor.go]: I also prefer the defer way. Since we have the mutex, even 1 timeout of 2s would block any other rules getting evaluated during that time which concerned me.

I could get rid of the wait group, which should be ok since all the healthchecks should be done after at most 2s, and cleanup period is 15s by default.

**Comment by tomwilkie** [comment]
> If you or someone else have any ideas

No ideas, but we had something similar with Bigtable; I just added a timeout and it unstuck it.  Are the ingester timeouts working okay?

>is anything we could do in this PR to make it easier to heartbeat with respect to #157? 

The two systems are quite decoupled right now; lets get #279 merged and see what shakes out.

**Comment by csmarchbanks** [comment]
Sounds good on waiting for #279. The ingester timeouts seem to be working ok, but every request will start timing out after 2s (push) or 10s (query), and that ingester connection never manages to correct itself.

**Comment by tomwilkie** [comment]
> The ingester timeouts seem to be working ok, but every request will start timing out after 2s (push) or 10s (query), and that ingester connection never manages to correct itself.

Is it worth destroying clients that experience a timeout then?

File: pkg/distributor/distributor.go (hunk 3 above)
Comment (tomwilkie) [reply pkg/distributor/distributor.go]: Of course (d'oh).

How about making it all nice and inline, synchronous code that builds a new clients dict without holding the lock, and then replaces the old one under the lock?

**Comment by tomwilkie** [comment]
Yeah, just as a thought...

**Comment by csmarchbanks** [comment]
It would work, and was something I was considering, but liked the healthcheck a bit more. I am happy to rework it based on what everyone thinks though.

Pros:
* No extra endpoint
* Would find an error on the first time, rather than waiting for up to 15s

Cons:
* Any future endpoints on the ingester would need the same handling, this could be abstracted but there would still be some entry function needed.
* Putting it in the maintenance loop keeps the maintenance code together
* Less straightforward to implement something like "restart if healthchecks fail for `x` time or `y` number of requests"
* Long queries/pushes could cause a false positive (seems unlikely though)

File: pkg/distributor/distributor.go (hunk 3 above)
Comment (csmarchbanks) [reply pkg/distributor/distributor.go]: I split the logic for removing stale ingester clients and healthchecking the ingester clients. I think everything is properly deferred now. Let me know what you think!

**Comment by tomwilkie** [comment]
You had me at "Any future endpoints on the ingester..."

File: pkg/distributor/distributor.go [hunk 4]
```diff
@@ -28,6 +28,7 @@ import (
 	ingester_client "github.com/weaveworks/cortex/pkg/ingester/client"
 	"github.com/weaveworks/cortex/pkg/ring"
 	"github.com/weaveworks/cortex/pkg/util"
+	"google.golang.org/grpc/health/grpc_health_v1"
```
Comment (tomwilkie) [inline pkg/distributor/distributor.go]: Can you move this import up to the block above please?  The pattern ~is~ should be [std lib import, 3rd party imports, repo local import].  I thought this was standard (as described here https://github.com/golang/go/wiki/CodeReviewComments#imports) but its turns out it might just be me being picky.

File: pkg/distributor/distributor.go [hunk 5]
```diff
@@ -201,14 +203,47 @@ func (d *Distributor) removeStaleIngesterClients() {
...
+	for _, ingester := range ingesters {
+		d.healthCheckAndRemoveIngester(ingester)
+	}
+}
+
+func (d *Distributor) healthCheckAndRemoveIngester(ingester *ring.IngesterDesc) {
+	client, err := d.getClientFor(ingester)
+	if err != nil {
+		d.removeClientFor(ingester, err)
+	}
```
Comment (tomwilkie) [inline pkg/distributor/distributor.go]: if err != nil, client == nil and the rest of this function will panic.

Comment (tomwilkie) [reply pkg/distributor/distributor.go]: Yeah I think it err != nil you don't need to remove the client either, you should just log and return (or return an error).

File: pkg/distributor/distributor.go [hunk 6]
```diff
@@ -201,14 +203,47 @@ func (d *Distributor) removeStaleIngesterClients() {
...
+
+	ctx, cancel := context.WithTimeout(context.Background(), d.cfg.RemoteTimeout)
+	defer cancel()
+	ctx = user.InjectOrgID(ctx, "0")
+
+	resp, err := client.Check(ctx, &grpc_health_v1.HealthCheckRequest{})
+	if err != nil || resp.Status != grpc_health_v1.HealthCheckResponse_SERVING {
```
Comment (tomwilkie) [inline pkg/distributor/distributor.go]: Please either return the error (and log it in the calling function) or log it here.  Slight preference for returning it and logging it in the caller.

**Comment by tomwilkie** [comment]
I think its looking much easier to follow wrt locking and concurrency.  A couple of comments, then we should be good to go.

It might also be worth extracting all this to a separate file (and even package), as there is starting to be enough logic to justify it.

**Comment by tomwilkie** [comment]
Yeah, the more I think about it the more I like the idea of putting it in `pkg/ingester/client` as a "client cache".

**Comment by tomwilkie** [comment]
Also, would you mind enabling CircleCI on your fork so we get test runs?  It should be free.

**Comment by bboreham** [comment]
I've done a pull/push to weaveworks/cortex so we get one test run now.

**Comment by csmarchbanks** [comment]
I did the refactor, and definitely think it cleans up distributor quite a bit. I also moved utils/compat to the ingester/client package since otherwise ingester/client could not import utils due to a circular dependency. 

Right now the loops for cleaning up stale clients, and health checking each client remain in distributor.go in order to not complicate what should just be a cache. If you want me to move those into the client package I can do that.

**Comment by csmarchbanks** [comment]
Not sure what's wrong with my circle ci yet... hopefully I will get some time to look at it tonight/tomorrow morning.

**Comment by csmarchbanks** [comment]
@bboreham or @tomwilkie Another review would be great when you get a chance!

File: pkg/distributor/distributor.go [hunk 7]
```diff
@@ -93,6 +93,7 @@ func (cfg *Config) RegisterFlags(f *flag.FlagSet) {
 	flag.DurationVar(&cfg.ClientCleanupPeriod, "distributor.client-cleanup-period", 15*time.Second, "How frequently to clean up clients for ingesters that have gone away.")
 	flag.Float64Var(&cfg.IngestionRateLimit, "distributor.ingestion-rate-limit", 25000, "Per-user ingestion rate limit in samples per second.")
 	flag.IntVar(&cfg.IngestionBurstSize, "distributor.ingestion-burst-size", 50000, "Per-user allowed ingestion burst size (in number of samples).")
+	flag.BoolVar(&cfg.HealthCheckIngesters, "distributor.health-check-ingesters", false, "Run a health check on each ingester client during the cleanup period.")
```
Comment (bboreham) [inline pkg/distributor/distributor.go]: can we say "during periodic cleanup" ?  First off I thought "the cleanup period" was maybe a period during shutdown when we clean up.

File: pkg/ingester/client/cache.go [hunk 8]
```diff
@@ -0,0 +1,89 @@
...
+
+	"github.com/go-kit/kit/log/level"
+	"github.com/weaveworks/cortex/pkg/util"
+)
+
+// Factory defines the signature for an ingester client factory
+type Factory func(addr string, cfg Config) (IngesterClient, error)
+
+// IngesterClientCache holds a cache of ingester clients
+type IngesterClientCache struct {
```
Comment (bboreham) [inline pkg/ingester/client/cache.go]: Is "cache" the right word?  Elsewhere I've seen this called a "connection pool", except we have max one connection per endpoint.  Maybe just some more explanation of the intended uses would help.

File: pkg/distributor/distributor.go [hunk 9]
```diff
@@ -184,52 +188,34 @@ func (d *Distributor) Stop() {
...
-	d.clientsMtx.Lock()
-	defer d.clientsMtx.Unlock()
-
 	ingesters := map[string]struct{}{}
 	for _, ing := range d.ring.GetAll() {
 		ingesters[ing.Addr] = struct{}{}
 	}
 
-	for addr, client := range d.clients {
+	for _, addr := range d.clientCache.RegisteredAddresses() {
```
Comment (bboreham) [inline pkg/distributor/distributor.go]: Seems we are doing two kinds of cleanup now: removing cache entries which are "stale" because the ring no longer references an ingester at that address, and removing entries which fail healthcheck.  Do we still need the first one?

File: pkg/distributor/distributor.go [hunk 10]
```diff
@@ -184,52 +188,34 @@ func (d *Distributor) Stop() {
...
-		return client, nil
-	}
-
-	client, err := d.cfg.ingesterClientFactory(ingester.Addr, d.cfg.IngesterClientConfig)
-	if err != nil {
-		return nil, err
+func (d *Distributor) healthCheckAndRemoveIngesters() {
+	for _, addr := range d.clientCache.RegisteredAddresses() {
+		client, err := d.clientCache.GetClientFor(addr)
+		if err != nil {
```
Comment (bboreham) [inline pkg/distributor/distributor.go]: This can only happen due to some race between `RegisteredAddresses()` and `GetClientFor()` - maybe change the cache API to remove the possibility?  Or move this whole loop into the cache?

File: pkg/ingester/client/cache.go (hunk 8 above)
Comment (csmarchbanks) [reply pkg/ingester/client/cache.go]: I like IngesterPool, I will change it to that, and add some more explanations.

File: pkg/distributor/distributor.go (hunk 9 above)
Comment (csmarchbanks) [reply pkg/distributor/distributor.go]: We could get away with only the healthcheck loop (after everyone has upgraded their ingesters and turned the flag on). However, I like the difference in expected behavior/logging from removing stale ingesters (expected) vs. failing ingesters (unexpected).

File: pkg/distributor/distributor.go (hunk 10 above)
Comment (csmarchbanks) [reply pkg/distributor/distributor.go]: I refactored this to the pool, but it still has an if around it in case someone deletes the entry from the pool while a previous healthcheck is happening. Right now that shouldn't happen, but that would be an annoying bug to run into if things change.

**Comment by csmarchbanks** [comment]
PR description updated

File: pkg/ingester/client/pool_test.go [hunk 11]
```diff
@@ -0,0 +1,125 @@
...
+	}
+
+	pool.GetClientFor("1")
+	if buildCount != 1 {
+		t.Errorf("Created client that should have been cached")
+	}
+
+	pool.GetClientFor("2")
+	if pool.Count() != 2 {
+		t.Errorf("Expected Count() = 2, got %d", pool.Count())
```
Comment (bboreham) [inline pkg/ingester/client/pool_test.go]: This is a bit wordy - we use `testify/assert` elsewhere to reduce to `assert.Equal(t, 2, pool.Count())`
